modules/networking/default.nix

```diff
@@ -7,24 +7,92 @@ let
 
   hostnameRegEx = ''^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$'';
 
-  emptyList = lst: if lst != [] then lst else ["empty"];
+  emptyList = lst: if lst != [ ] then lst else [ "empty" ];
 
   onOff = cond: if cond then "on" else "off";
 
-  setNetworkServices = optionalString (cfg.knownNetworkServices != []) ''
-    networkservices=$(networksetup -listallnetworkservices)
-    ${concatMapStringsSep "\n" (srv: ''
-      case "$networkservices" in
-        *${lib.escapeShellArg srv}*)
-          networksetup -setdnsservers ${lib.escapeShellArgs ([ srv ] ++ (emptyList cfg.dns))}
-          networksetup -setsearchdomains ${lib.escapeShellArgs ([ srv ] ++ (emptyList cfg.search))}
-          ;;
-      esac
-    '') cfg.knownNetworkServices}
+  setLocations = optionalString (cfg.knownNetworkServices != [ ] && cfg.location != { }) ''
+    curr_location=$(networksetup -getcurrentlocation)
+
+    readarray -t curr_locations_array < <(networksetup -listlocations)
+
+    declare -A curr_locations
+    for location in "''${curr_locations_array[@]}"; do
+      curr_locations[$location]=1
+    done
+
+    declare -A goal_locations
+    for location in ${strings.escapeShellArgs (builtins.attrNames cfg.location)}; do
+      goal_locations[$location]=1
+    done
+
+    for location in "''${!goal_locations[@]}"; do
+      if [[ ! -v curr_locations[$location] ]]; then
+        networksetup -createlocation "$location" populate > /dev/null
+      fi
+    done
+
+    # switch to a location that surely does not need to be deleted
+    networksetup -switchtolocation ${strings.escapeShellArg (builtins.head (builtins.attrNames cfg.location))} > /dev/null
+
+    for location in "''${!curr_locations[@]}"; do
+      if [[ ! -v goal_locations[$location] ]]; then
+        networksetup -deletelocation "$location" > /dev/null
+      fi
+    done
```

Comment:
This is going to clobber unmanaged locations, right? That’s usually why we have `known*` stuff. That might be okay, although it’s unfortunate that going from *n* to 0 locations won’t be handled.